cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -381,36 +432,51 @@ void recordProgramStopOldFormat(ProgramId programId, String pid, long stopTs, Pr
 
   private void recordProgramStop(ProgramId programId, String pid, long stopTs, ProgramRunStatus runStatus,
                                  @Nullable BasicThrowable failureCause, MDSKey.Builder builder) {
-    MDSKey key = getProgramKeyBuilder(TYPE_RUN_RECORD_STARTED, programId)
+    MDSKey key = getProgramKeyBuilder(TYPE_RUN_RECORD_RUNNING, programId)
       .add(pid)
       .build();
-    RunRecordMeta started = getFirst(key, RunRecordMeta.class);
+    RunRecordMeta existing = getFirst(key, RunRecordMeta.class);
 
     // Check without the version string only for default version
-    if (!upgradeComplete.get() && started == null && (programId.getVersion().equals(ApplicationId.DEFAULT_VERSION))) {
-      key = getVersionLessProgramKeyBuilder(TYPE_RUN_RECORD_STARTED, programId)
+    if (!upgradeComplete.get() && existing == null && (programId.getVersion().equals(ApplicationId.DEFAULT_VERSION))) {
+      key = getVersionLessProgramKeyBuilder(TYPE_RUN_RECORD_RUNNING, programId)
         .add(pid)
         .build();
-      started = getFirst(key, RunRecordMeta.class);
+      existing = getFirst(key, RunRecordMeta.class);
     }
 
-    if (started == null) {
-      String msg = String.format("No meta for started run record for namespace %s app %s version %s program type %s " +
-                                 "program %s pid %s exists",
-                                 programId.getNamespace(), programId.getApplication(), programId.getVersion(),
-                                 programId.getType().name(), programId.getProgram(), pid);
-      LOG.error(msg);
-      throw new IllegalArgumentException(msg);
+    if (existing == null) {
+      // No running record, so program could have started, but encountered an error before reaching the running state
+      key = getProgramKeyBuilder(TYPE_RUN_RECORD_STARTED, programId)
```

Comment:
I'm not sure why `TYPE_RUN_RECORD_SUSPENDED` was not checked in previous version though. Maybe try the test before you make the change and see what is the result.